From 305d2ed186c9e1e7c3d0f914235e926e27b35a9f Mon Sep 17 00:00:00 2001
From: Miklos Vajna <vmiklos@collabora.co.uk>
Date: Tue, 15 Dec 2020 17:28:49 +0000
Subject: [PATCH] Add FPDFAnnot_GetBorder() API

This is similar to FPDFAnnot_GetLine(), but a /Border key is valid for
all annotation types.

The array can have an optional 4th element, but that itself doesn't have
a fixed size, so it's not exposed by this API -- could be done by a
separate function.

Change-Id: I796c2ea17be1f039a6c6129dbde0b0a22276e19a
Reviewed-on: https://pdfium-review.googlesource.com/c/pdfium/+/77010
Commit-Queue: Tom Sepez <tsepez@chromium.org>
Reviewed-by: Tom Sepez <tsepez@chromium.org>
---
 fpdfsdk/fpdf_annot.cpp              | 23 +++++++++++++++++
 fpdfsdk/fpdf_annot_embeddertest.cpp | 40 +++++++++++++++++++++++++++++
 fpdfsdk/fpdf_view_c_api_test.c      |  1 +
 public/fpdf_annot.h                 | 16 ++++++++++++
 testing/resources/line_annot.in     |  2 ++
 testing/resources/line_annot.pdf    |  6 +++--
 6 files changed, 86 insertions(+), 2 deletions(-)

diff --git a/fpdfsdk/fpdf_annot.cpp b/fpdfsdk/fpdf_annot.cpp
index 85f86e547..eee3567d4 100644
--- a/fpdfsdk/fpdf_annot.cpp
+++ b/fpdfsdk/fpdf_annot.cpp
@@ -913,6 +913,29 @@ FPDF_EXPORT FPDF_BOOL FPDF_CALLCONV FPDFAnnot_GetLine(FPDF_ANNOTATION annot,
   return true;
 }
 
+FPDF_EXPORT FPDF_BOOL FPDF_CALLCONV
+FPDFAnnot_GetBorder(FPDF_ANNOTATION annot,
+                    float* horizontal_radius,
+                    float* vertical_radius,
+                    float* border_width) {
+  if (!horizontal_radius || !vertical_radius || !border_width)
+    return false;
+
+  CPDF_Dictionary* annot_dict = GetAnnotDictFromFPDFAnnotation(annot);
+  if (!annot_dict)
+    return false;
+
+  CPDF_Array* border = annot_dict->GetArrayFor(pdfium::annotation::kBorder);
+  if (!border || border->size() < 3)
+    return false;
+
+  *horizontal_radius = border->GetNumberAt(0);
+  *vertical_radius = border->GetNumberAt(1);
+  *border_width = border->GetNumberAt(2);
+
+  return true;
+}
+
 FPDF_EXPORT FPDF_BOOL FPDF_CALLCONV FPDFAnnot_HasKey(FPDF_ANNOTATION annot,
                                                      FPDF_BYTESTRING key) {
   CPDF_Dictionary* pAnnotDict = GetAnnotDictFromFPDFAnnotation(annot);
diff --git a/public/fpdf_annot.h b/public/fpdf_annot.h
index 6c8a237e7..2b708af13 100644
--- a/public/fpdf_annot.h
+++ b/public/fpdf_annot.h
@@ -452,6 +452,22 @@ FPDF_EXPORT FPDF_BOOL FPDF_CALLCONV FPDFAnnot_GetLine(FPDF_ANNOTATION annot,
                                                       FS_POINTF* start,
                                                       FS_POINTF* end);
 
+// Experimental API.
+// Get the characteristics of the annotation's border (rounded rectangle).
+//
+//   annot  - handle to an annotation, as returned by e.g. FPDFPage_GetAnnot()
+//   horizontal_radius - horizontal corner radius, in default user space units
+//   vertical_radius - vertical corner radius, in default user space units
+//   border_width - border width, in default user space units
+//
+// Returns true if |horizontal_radius|, |vertical_radius| and |border_width| are
+// not NULL, false otherwise.
+FPDF_EXPORT FPDF_BOOL FPDF_CALLCONV
+FPDFAnnot_GetBorder(FPDF_ANNOTATION annot,
+                    float* horizontal_radius,
+                    float* vertical_radius,
+                    float* border_width);
+
 // Experimental API.
 // Check if |annot|'s dictionary has |key| as a key.
 //
-- 
2.26.2

